src/lib/storage/column_iterables.hpp

```diff
@@ -80,8 +82,25 @@ class ColumnIterable {
     });
   }
 
+  /**
+   * By default, iterables may call functors passed to with_iterators()
+   * several times with iterators to different parts of the column.
+   * Call require_single_functor_call() if your code requires a single call.
+   */
+  void require_single_functor_call() { _single_call_required = true; }
```

Comment:
I don't understand this part, could you give an example?

Reply:
Yes. This method is required for the `ReferenceColumnIterable`. The old implementation wasn’t ideal at all because of two reasons. Firstly the iterable duplicated the code needed to access underlying columns, which wasn’t so much a problem because there were only two possibilities. Now their much more. Secondly, the iterator had to resolve the type of the referenced column for each element. If the reference column references only a single column, this is a lot of wasted performance. But it even if the reference column references several columns, it makes more sense to check in advance how many elements reference the same column and resolve the type once, but this wasn’t possibly due to the separation of the increment and the dereference operation of the iterator. So now, the `ReferenceColumnIterable` looks for runs uses the iterables of the other columns, which solves both problems. But this is only possible, if the functor passed into `with_iterators` can be called multiple times, once for each run. There is no difference for `for_each` however. The only situation where the functor cannot be called multiple times is when iterating over two columns simultaneously which is currently done in `nested_loop_join.cpp` and `column_comparison_table_scan_impl.cpp`.

Reply:
It is used here: https://github.com/hyrise/hyrise/blob/ee1d70795342a92182335be70c66e4122f47e3e0/src/lib/storage/reference_column/reference_column_iterable.hpp

Reply:
Got it. I wonder if a name in the direction of `resolve_type_per_row` would make this clearer.

Reply:
`resolve_type_per_row` is not much better I think :/

Reply:
Ok, maybe you can incorporate part of your comment above into the comment in the code.